File: src/main/java/JvmMon.java
import com.eclipsesource.v8.*;
import com.eclipsesource.v8.utils.V8ObjectUtils;
import org.andrejs.json.Json;

import java.io.IOError;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

public class JvmMon implements JavaCallback {

    private static NodeJS nodeJS = NodeJS.createNodeJS();
    private static V8 v8 = nodeJS.getRuntime();

    private JvmStats jvmStats = new JvmStats();

    public static void main(String[] args) throws Exception {

        Path script = Paths.get("jvm-mon.js");
        if(!Files.exists(script)) {
            script = Paths.get("src", "dist", "jvm-mon.js");
        }

        JvmMon jvmMon = new JvmMon();

        v8.registerJavaMethod(jvmMon, "getJvmStats");
        nodeJS.exec(script.toFile());

        while(nodeJS.isRunning()) {
            nodeJS.handleMessage();
        }

        nodeJS.release();
    }

    public JvmMon() {
    }

    /** Called from javascript to get new stats **/
    @Override
    public Object invoke(V8Object receiver, V8Array parameters) {
        try {
            int selectedVmId = parameters.getInteger(0);
            List<Json> vmStats = jvmStats.getVmStats(selectedVmId);
            return V8ObjectUtils.toV8Array(v8, vmStats);

        } catch (Exception e) {
            throw new IOError(e);
        }
    }
}

File: src/main/kotlin/JvmMon.kt
import com.eclipsesource.v8.JavaCallback
import com.eclipsesource.v8.NodeJS
import com.eclipsesource.v8.V8Array
import com.eclipsesource.v8.V8Object
import com.eclipsesource.v8.utils.V8ObjectUtils
import jvmmon.JvmStats
import java.io.File
import java.io.File.pathSeparator

class JvmMon : JavaCallback {

    private val jvmStats = JvmStats()

    /** Called from node to get new stats */
    override fun invoke(receiver: V8Object, parameters: V8Array): Any {
        val selectedVmId = parameters.getInteger(0)
        val vmStats = jvmStats.getVmStats(selectedVmId)

        return V8ObjectUtils.toV8Array(v8, vmStats)
    }

    companion object {

        private val nodeJS = NodeJS.createNodeJS()
        private val v8 = nodeJS.runtime

        @JvmStatic fun main(args: Array<String>) {
            v8.registerJavaMethod(JvmMon(), "getJvmStats")

            runNodeScript(findScript())
        }

        private fun runNodeScript(script: File) {
            nodeJS.exec(script)

            while (nodeJS.isRunning) {
                nodeJS.handleMessage()
            }

            nodeJS.release()
        }

        private fun findScript(): File {
            val script = File("jvm-mon.js")

            return if (script.exists())
                return script
            else {
                File("src${pathSeparator}dist", "jvm-mon.js")
            }

        }
    }
}
